packages/browser-repl/src/components/shell.tsx

```diff
@@ -253,235 +292,181 @@ export class _Shell extends Component<ShellProps, ShellState> {
-    );
-    history.splice(this.props.maxHistoryLength);
-
-    Object.freeze(history);
-
-    return history;
-  }
-
-  private addEntriesToOutput(
-    entries: readonly ShellOutputEntry[]
-  ): readonly ShellOutputEntry[] {
-    const output = [...this.state.output, ...entries];
-
-    output.splice(0, output.length - this.props.maxOutputLength);
-
-    Object.freeze(output);
-
-    return output;
-  }
-
-  onClearCommand = (): void => {
-    const output: [] = [];
-
-    Object.freeze(output);
-
-    this.setState({ output });
-    this.props.onOutputChanged(output);
-  };
-
-  onPrint = (result: { type: string | null; printable: any }[]): void => {
-    const output = this.addEntriesToOutput(
-      result.map((entry) => ({
-        format: 'output',
-        type: entry.type,
-        value: entry.printable,
-      }))
-    );
-    this.setState({ output });
-    this.props.onOutputChanged(output);
-  };
-
-  onPrompt = (
-    question: string,
-    type: 'password' | 'yesno'
-  ): Promise<string> => {
-    if (type !== 'password') {
-      return Promise.reject(new Error('yes/no prompts not implemented yet'));
-    }
-    const reset = () => {
-      this.onFinishPasswordPrompt = noop;
-      this.onCancelPasswordPrompt = noop;
-      this.setState({ passwordPrompt: '' });
-      setTimeout(this.focusEditor, 1);
-    };
-
-    const ret = new Promise<string>((resolve, reject) => {
-      this.onFinishPasswordPrompt = (result: string) => {
-        reset();
-        resolve(result);
-      };
-      this.onCancelPasswordPrompt = () => {
-        reset();
-        reject(new Error('Canceled by user'));
-      };
-    });
-    this.setState({ passwordPrompt: question });
-    return ret;
-  };
-
-  private onInput = async (code: string): Promise<void> => {
-    if (!code || code.trim() === '') {
-      this.appendEmptyInput();
-      return;
-    }
-
-    const inputLine: ShellOutputEntry = {
-      format: 'input',
-      value: code,
-    };
+      try {
+        onOperationStarted?.();
+
+        runtime.setEvaluationListener(listener);
+        const result = await runtime.evaluate(code);
+        outputLine = {
+          format: 'output',
+          type: result.type,
+          value: result.printable,
+        };
+      } catch (error) {
+        outputLine = {
+          format: 'error',
+          value: error,
+        };
+      } finally {
+        await updateShellPrompt();
+        onOperationEnd?.();
+      }
 
-    let output = this.addEntriesToOutput([inputLine]);
-    this.setState({
-      operationInProgress: true,
-      output,
-    });
-    this.props.onOutputChanged(output);
+      return outputLine;
+    },
+    [listener, onOperationEnd, onOperationStarted, runtime, updateShellPrompt]
+  );
 
-    const outputLine = await this.evaluate(code);
+  const onInput = useCallback(
+    async (code: string) => {
+      const newOutput = (output ?? []).slice();
+      const newHistory = (history ?? []).slice();
```

Comment:
nit: usually copying is a bit easier to read with object/array spread imo:

```suggestion
      const newOutput = [...output ?? []];
      const newHistory = [...history ?? []];
```